pingcap/tidb-operator
# retrieve region leader from tikv

### What problem does this PR solve?

Fix #3781
### What is changed and how does it work?

Root cause:

Currently, we retrieve the region count leader from PD to determine whether the leaders of one TiKV have been transferred to the other ones, however, if one PD is restarted and then becomes the leader in 1 minute, it may not have all the region leader info for each TiKV, so we have to change the logic to retrieve the region leader from each TiKV directly.
### Code changes


### Tests


  - Deploy a TidbCluster with 3 PDs and 3 TiKVs, do not enable TLS, run sysbench to the cluster, upgrade PD and TiKV, check that the QPS does not drop obviously during the TiKV upgrade.
```
[ 140s ] thds: 16 tps: 179.20 qps: 3574.92 (r/w/o: 2500.91/715.60/358.40) lat (ms,95%): 125.52 err/s: 0.00 reconn/s: 0.00
[ 150s ] thds: 16 tps: 174.70 qps: 3503.41 (r/w/o: 2453.31/700.60/349.50) lat (ms,95%): 130.13 err/s: 0.10 reconn/s: 0.00
[ 160s ] thds: 16 tps: 187.91 qps: 3759.09 (r/w/o: 2632.73/750.44/375.92) lat (ms,95%): 116.80 err/s: 0.10 reconn/s: 0.00
[ 170s ] thds: 16 tps: 179.68 qps: 3598.97 (r/w/o: 2519.50/720.11/359.36) lat (ms,95%): 125.52 err/s: 0.10 reconn/s: 0.00
[ 180s ] thds: 16 tps: 179.20 qps: 3580.78 (r/w/o: 2505.09/717.20/358.50) lat (ms,95%): 130.13 err/s: 0.00 reconn/s: 0.00
[ 190s ] thds: 16 tps: 169.10 qps: 3382.45 (r/w/o: 2369.67/674.59/338.20) lat (ms,95%): 134.90 err/s: 0.00 reconn/s: 0.00
[ 200s ] thds: 16 tps: 172.00 qps: 3440.76 (r/w/o: 2408.64/688.11/344.01) lat (ms,95%): 132.49 err/s: 0.00 reconn/s: 0.00
[ 210s ] thds: 16 tps: 176.40 qps: 3529.84 (r/w/o: 2469.46/707.59/352.79) lat (ms,95%): 125.52 err/s: 0.00 reconn/s: 0.00
[ 220s ] thds: 16 tps: 180.40 qps: 3598.26 (r/w/o: 2518.24/719.41/360.61) lat (ms,95%): 127.81 err/s: 0.00 reconn/s: 0.00
```
  - Deploy a TidbCluster with 3 PDs and 3 TiKVs, enable TLS, run sysbench to the cluster, upgrade PD and TiKV, check that the QPS does not drop obviously during the TiKV upgrade.
```
[ 140s ] thds: 16 tps: 184.00 qps: 3679.82 (r/w/o: 2576.62/735.10/368.10) lat (ms,95%): 125.52 err/s: 0.10 reconn/s: 0.00
[ 150s ] thds: 16 tps: 178.80 qps: 3577.67 (r/w/o: 2502.88/717.19/357.60) lat (ms,95%): 130.13 err/s: 0.00 reconn/s: 0.00
[ 160s ] thds: 16 tps: 187.20 qps: 3741.21 (r/w/o: 2619.71/746.90/374.60) lat (ms,95%): 127.81 err/s: 0.20 reconn/s: 0.00
[ 170s ] thds: 16 tps: 173.10 qps: 3462.26 (r/w/o: 2422.57/693.39/346.30) lat (ms,95%): 130.13 err/s: 0.10 reconn/s: 0.00
[ 180s ] thds: 16 tps: 169.09 qps: 3382.90 (r/w/o: 2367.83/676.88/338.19) lat (ms,95%): 134.90 err/s: 0.00 reconn/s: 0.00
[ 190s ] thds: 16 tps: 169.81 qps: 3407.72 (r/w/o: 2386.99/681.02/339.71) lat (ms,95%): 134.90 err/s: 0.10 reconn/s: 0.00
[ 200s ] thds: 16 tps: 162.80 qps: 3249.27 (r/w/o: 2274.78/648.89/325.60) lat (ms,95%): 144.97 err/s: 0.00 reconn/s: 0.00
[ 210s ] thds: 16 tps: 164.20 qps: 3286.95 (r/w/o: 2301.53/657.01/328.40) lat (ms,95%): 144.97 err/s: 0.00 reconn/s: 0.00
[ 220s ] thds: 16 tps: 159.69 qps: 3191.06 (r/w/o: 2232.70/638.97/319.39) lat (ms,95%): 144.97 err/s: 0.00 reconn/s: 0.00
```

### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Retrieve the region leader count from TiKV Pod directly instead of from PD to get the accurate count
```
